src/ht.c

```diff
@@ -1,239 +1,184 @@
+// Inserts the key value pair key, value into the hashtable.
+// if update is true, allow for updates, otherwise return false if the key
+// already exists.
+static bool internal_ht_insert(SdbHash* ht, bool update, const char* key,
+		const char* value) {
+	SdbKv* kvp;
+	ut64 hash;
+	ut64 bucket;
+	bool found = true;
+	if (update) {
+		ht_delete (ht, key);
+	} else {
+		(void)ht_find (ht, key, &found);
+	}
+	if (update || !found) {
+		kvp = calloc (1, sizeof (SdbKv *));
+		if (kvp) {
+			if (ht->dupkey) {
+				kvp->key = ht->dupkey (key);
+			} else {
+				kvp->key = (char*)key;
+			}
+			if (ht->dupvalue) {
+				kvp->value = ht->dupvalue (value);
+			} else {
+				kvp->value = (char*)value;
+			}
+			hash = ht->hashfn (key);
+			bucket = hash % ht->size;
+			kvp->expire = 0;
+			kvp->value_len = strlen (kvp->value);
+			ls_prepend (ht->table[bucket], kvp);
+			ht->count++;
+			// Check if we need to grow the table.
+			if (ht->count >= ht->load_factor * ht->size) {
+				internal_ht_grow (ht);
+			}
+			return true;
+		}
+	}
+	return false;
 }
 
-#if 0
-void ht_set(SdbHash *ht, ut32 hash, void *data) {
-	SdbHashEntry *e = ht_search (ht, hash);
-	if (e) {
-		if (ht->list->free)
-			ht->list->free (e->data);
-		e->data = data;
-		e->iter->data = data;
-	} else ht_insert (ht, hash, data, NULL);
+// Inserts the key value pair key, value into the hashtable.
+// Doesn't allow for "update" of the value.
+bool ht_insert (SdbHash* ht, const char* key, const char* value) {
+	return internal_ht_insert (ht, false, key, value);
 }
-#endif
-
-/**
- * Inserts the data with the given hash into the table.
- *
- * Note that insertion may rearrange the table on a resize or rehash,
- * so previously found hash_entries are no longer valid after this function.
- */
-int ht_insert(SdbHash *ht, ut32 hash, void *data, SdbListIter *iter) {
-	ut32 hash_address;
-	if (!ht || !data)
-		return 0;
 
-	if (ht->entries >= ht->max_entries)
-		ht_rehash (ht, ht->size_index + 1);
-	else if (ht->deleted_entries + ht->entries >= ht->max_entries)
-		ht_rehash (ht, ht->size_index);
-
-	hash_address = hash % ht->size;
-	do {
-		SdbHashEntry *entry = ht->table + hash_address;
-		ut32 double_hash;
+// Inserts the key value pair key, value into the hashtable.
+// Does allow for "update" of the value.
+bool ht_update (SdbHash* ht, const char* key, const char* value) {
+	return internal_ht_insert (ht, true, key, value);
+}
 
-		if (!entry_is_present (entry)) {
-			if (entry_is_deleted (entry))
-				ht->deleted_entries--;
-			entry->hash = hash;
-			entry->data = data;
-			entry->iter = rehash? iter: ls_append (ht->list, data);
-			ht->entries++;
-			return 1;
+// Looks up the corresponding value from the key. Returns true if found, false
+// otherwise.
+SdbKv* ht_find_kvp(SdbHash* ht, const char* key, bool* found) {
+	ut64 hash;
+	ut64 bucket;
+	SdbListIter* iter;
+	SdbKv* kvp;
+	hash = ht->hashfn (key);
+	bucket = hash % ht->size;
+	ls_foreach (ht->table[bucket], iter, kvp) {
```

Comment:
C# ftw

> On 23 Jun 2016, at 22:32, Jeffrey Crowell <notifications@github.com> wrote:
> 
> In src/ht.c:
> 
> > -				ht->deleted_entries--;
> > -			entry->hash = hash;
> > -			entry->data = data;
> > -			entry->iter = rehash? iter: ls_append (ht->list, data);
> > -			ht->entries++;
> > -			return 1;
> > +// Looks up the corresponding value from the key. Returns true if found, false
> > +// otherwise.
> > +SdbKv* ht_find_kvp(SdbHash* ht, const char* key, bool* found) {
> > +	ut64 hash;
> > +	ut64 bucket;
> > +	SdbListIter* iter;
> > +	SdbKv* kvp;
> > +	hash = ht->hashfn (key);
> > +	bucket = hash % ht->size;
> > +	ls_foreach (ht->table[bucket], iter, kvp) {
> now uses the sizes used in c#'s hashtable impl.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>